Changed evaluation of 'engine' to sustain sprintf format (#25)
logstash-plugins/logstash-output-elastic_app_search: docs/index.asciidoc

  * There is no default value

The Engine name. Engine is your search engine created in App Search, an information repository that includes the indexed document records.

> Engine field also support `sprintf` format, so a valid name could be `engine-%{eng_level}`, in this case the engine

Reviewer: Maybe link to `sprintf` - `{logstash-ref}/event-dependent-configuration.html#sprintf` docs for a usage example
```suggestion
Engine field also support {logstash-ref}/event-dependent-configuration.html#sprintf[sprintf format], so a valid name could be `engine-%{eng_level}`, in this case the engine
```

> value is resolved at runtime, event by event; dynamic resolution could lead to invalid engine names in which case

Reviewer: Maybe something like:

Engine field also support `sprintf` format to enable the engine name to be derived from a field value from each event, for  example `engine-%{engine_name}`. 

Caution: Invalid engine names will cause ingestion to be halted until the field value can be resolved into a valid engine name. This be caused by either the field value being interpolated resolving to a value without a matching engine, or, if the field is missing from the event, to not be resolved at all. To catch errors where the field is missing from the event, it may be advisable to add a "default" engine type and handle it in the configuration, eg:

[source,ruby]
----------------------------------
input {
  stdin {
    codec => json
  }
}
filter {
  if ![engine_name] {
    mutate {
      add_field => {"engine_name" => "default"}
    }
  }
}
output {
  appsearch {
    engine => "engine_%{[engine_name]}"
  }
}

We could possibly extend the example to check the valid engine names as well as if it is missing and tag them appropriately, and at some point we might want to consider a follow up PR to do this inside the plugin

Author: Thanks for suggestion @robbavey, applied :-)

I've a question about:
> extend the example to check the valid engine names

if I understand correctly, you mean to do an HTTP call to the AppSearch service to validate the resolved engine name, this could double the calls to AppSearch, one for the check and one for effectively index the data. To avoid some calls the pipeline should provide some form of caching, keeping the already validated engine names. This could be done with `ruby` filter maybe. Is it what you mean?

Reviewer: Actually, my initial thought for an example was something simpler - just a manual check that the engine name is in a list of engine names that the user would provide in the configuration. Not a great solution in the long term, depending on how many engines a user would typically include in a cluster, and how often that list changes, but it may be of some help

  * There is no default value

> The Engine name. Engine is your search engine created in App Search, an information repository that includes the indexed document records.

Reviewer 2: ```suggestion
The name of the search engine you created in App Search, an information
repository that includes the indexed document records.
```

> Engine field also support {logstash-ref}/event-dependent-configuration.html#sprintf[sprintf format], to enable

Reviewer: Should `Engine field` link to the docs for that field?

Author: That could be nice, I'll add the link with the suggestions that come from @karenzone

Reviewer: Ignore this suggestion - I didn't notice that this was part of the definition for `engine`... 🤦

> Engine field also support {logstash-ref}/event-dependent-configuration.html#sprintf[sprintf format], to enable
> the engine name to be derived from a field value from each event, for example `engine-%{engine_name}`.
> 
> Caution:
> Invalid engine names will cause ingestion to be halted until the field value can be resolved into a valid engine name.
> This be caused by either the field value being interpolated resolving to a value without a matching engine,
> or, if the field is missing from the event, to not be resolved at all.
> To catch errors where the field is missing from the event, it may be advisable to add a "default" engine type
> and handle it in the configuration, eg:
> 

Reviewer 2: ```suggestion
The `engine` field supports
{logstash-ref}/event-dependent-configuration.html#sprintf[sprintf format] to
allow the engine name to be derived from a field value from each event, for
example `engine-%{engine_name}`.

Invalid engine names cause ingestion to stop until the field value can be resolved into a valid engine name.
This situation can happen if the interpolated field value resolves to a value without a matching engine,
or, if the field is missing from the event and cannot be resolved at all.

TIP: Consider adding a "default" engine type in the configuration to catch errors if
the field is missing from the event.

Example:

```
